File: fast_sync_dart_client/fast_sync_client/lib/src/implementation/data/syncable_object_repository.dart
import 'package:fast_sync_client/fast_sync_client.dart';

class SyncalbeRepository<T extends ISyncableObject>
    implements ISyncableRepository<T> {
  final ISyncableDataSource<T> dataSource;

  ISyncableDataSource<ISyncMetadata> get syncMetadataDataSource {
    return FastSync.getsyncMetadataDataSource();
  }

  SyncalbeRepository({required this.dataSource});

  @override
  Future<T> add(T entity) async {
    entity.metadata.syncOperation = SyncOperationEnum.add.code;
    entity.metadata.type = T.toString();
    entity.metadata.syncZone = FastSync.getTypeSyncZone(T.toString());
    entity = _dirtyObject(entity);
    entity = _linkMetadataId(entity, entity.metadata);
    await syncMetadataDataSource.add(entity.metadata);
    return await dataSource.add(entity);
  }

  @override
  Future<List<T>> addMany(List<T> entities, ISyncMetadata metadata) async {
    List<T> entitiesToSave = [];
    List<ISyncMetadata> metadataTosave = [];
    for (var entity in entities) {
      entity.metadata.syncOperation = SyncOperationEnum.add.code;
      entity.metadata.type = T.toString();
      entity.metadata.syncZone = FastSync.getTypeSyncZone(T.toString());
      entity = _dirtyObject(entity);
      entity = _linkMetadataId(entity, entity.metadata);
      entitiesToSave.add(entity);
      metadataTosave.add(entity.metadata);
    }
    await syncMetadataDataSource.addMany(metadataTosave);
    return await dataSource.addMany(entitiesToSave);
  }

  @override
  Future<int> count() async {
    return await dataSource.count();
  }

  @override
  void dispose() async {
    return await dataSource.dispose();
  }

  @override
  Future<T?> findById(String id) async {
    var item = await dataSource.findById(id);
    String metadataId = item?.metadataId as String;
    var metadata = await syncMetadataDataSource.findById(metadataId);
    if (item != null && metadata != null) {
      if (item.deleted) {
        return null;
      }
      item.metadata = metadata;
      return item;
    } else {
      return null;
    }
  }

  @override
  Future<List<T>> getAll() async {
    List<T> items = await dataSource.getAll();
    List<T> itemsToReturn = [];
    items = items.where(_undoRemovedEntities).toList();
    List<String> metadataIds = items.map((m) => m.metadataId).toList();
    List<ISyncMetadata> metadatas =
        await this.syncMetadataDataSource.findByIds(metadataIds);
    var metadataDict = createDict(metadatas);
    for (var entity in items) {
      entity.metadata = (metadataDict[entity.metadataId]) as ISyncMetadata;
      itemsToReturn.add(entity);
    }
    return itemsToReturn;
  }

  @override
  Future<List<T>> query(bool Function(T) query) async {
    List<T> items = await this.getAll();
    List<T> filteredItems = items.where(query).toList();
    return filteredItems;
  }

  @override
  Future<List<T>> removeMany(List<T> entities, ISyncMetadata metadata) async {
    List<T> entitiesToSave = [];
    List<ISyncMetadata> metadataTosave = [];
    for (var entity in entities) {
      entity.deleted = true;
      entity.metadata.syncOperation = SyncOperationEnum.delete.code;
      entity.metadata.type = T.toString();
      entity = _dirtyObject(entity);
      entity = _linkMetadataId(entity, entity.metadata);
      entitiesToSave.add(entity);
      metadataTosave.add(entity.metadata);
    }
    await syncMetadataDataSource.updateMany(metadataTosave);
    return await this.dataSource.updateMany(entitiesToSave);
  }

  @override
  Future<T> update(T entity) async {
    entity.metadata.syncOperation = SyncOperationEnum.update.code;
    entity.metadata.type = T.toString();
    entity = _dirtyObject(entity);
    entity = _linkMetadataId(entity, entity.metadata);
    await syncMetadataDataSource.update(entity.metadata.id, entity.metadata);
    return await dataSource.update(entity.id, entity);
  }

  @override
  Future<List<T>> updateMany(List<T> entities, ISyncMetadata metadata) async {
    List<T> entitiesToSave = [];
    List<ISyncMetadata> metadataTosave = [];

    for (var entity in entities) {
      entity.metadata.syncOperation = SyncOperationEnum.update.code;
      entity.metadata.type = T.toString();
      entity = _dirtyObject(entity);
      entity = _linkMetadataId(entity, entity.metadata);
      entitiesToSave.add(entity);
      metadataTosave.add(entity.metadata);
    }
    await syncMetadataDataSource.updateMany(metadataTosave);
    return await dataSource.updateMany(entitiesToSave);
  }

  @override
  Future<List> undirtyList(List entities) async {
    List entitiesToSave = [];
    List<ISyncMetadata> metadataTosave = [];
    for (var entity in entities) {
      entity.dirty = false;
      entity = _linkMetadataId(entity, entity.metadata);
      entitiesToSave.add(entity);
      metadataTosave.add(entity.metadata);
    }
    await dataSource.syncUpdate(entitiesToSave);
    await syncMetadataDataSource.syncUpdate(metadataTosave);
    return entitiesToSave;
  }

  @override
  Future<void> hardDelete() async {
    SyncVersionManager versionManager = FastSync.getSyncVersionManager();
    await dataSource.hardDelete();
    await versionManager.resetTypeSyncVersion(T.toString());
  }

  bool _undoRemovedEntities(T entity) {
    return !entity.deleted;
  }

  T _dirtyObject(T object) {
    if (!object.dirty) {
      object.dirty = true;
      object.metadata.version++;
    }
    return object;
  }

  T _linkMetadataId(T object, ISyncMetadata metadata) {
    object.metadataId = metadata.id;
    return object;
  }
}

File: fast-sync-typescript-client-sdk/packages/fast-sync-client/src/implementation/data/syncable_object_repository.ts
import { FastSync } from "../fast_sync";
import { ISyncMetadata } from "../../abstraction/metadata/isync_metadata";
import { SyncOperationEnum } from "../../abstraction/metadata/isync_operation";
import { ISyncableDataSource } from "../../abstraction/data/isyncable_data_source";
import { ISyncableRepository } from "../../abstraction/data/isyncable_repository";
import { ISyncableObject } from "../../abstraction/metadata/isyncable_object";

export class SyncalbeRepository<T extends ISyncableObject>
  implements ISyncableRepository<T>
{
  private dataSource: ISyncableDataSource<T>;

  constructor(dataSource: ISyncableDataSource<T>, private type: string) {
    this.dataSource = dataSource;
  }

  get syncMetadataDataSource(): ISyncableDataSource<ISyncMetadata> {
    return FastSync.getsyncMetadataDataSource();
  }

  async add(entity: T): Promise<T> {
    entity.metadata.syncOperation = SyncOperationEnum.add;
    entity.metadata.type = this.type;
    entity.metadata.syncZone = FastSync.getTypeSyncZone(this.type);
    entity = this._dirtyObject(entity);
    entity = this._linkMetadataId(entity, entity.metadata);
    await this.syncMetadataDataSource.add(entity.metadata);
    return await this.dataSource.add(entity);
  }

  async addMany(entities: T[]): Promise<T[]> {
    const entitiesToSave: T[] = [];
    const metadataTosave: ISyncMetadata[] = [];
    for (let entity of entities) {
      entity.metadata.syncOperation = SyncOperationEnum.add;
      entity.metadata.type = this.type;
      entity.metadata.syncZone = FastSync.getTypeSyncZone(this.type);
      entity = this._dirtyObject(entity);
      entity = this._linkMetadataId(entity, entity.metadata);
      entitiesToSave.push(entity);
      metadataTosave.push(entity.metadata);
    }

    await this.syncMetadataDataSource.addMany(metadataTosave);
    return await this.dataSource.addMany(entitiesToSave);
  }

  async count(): Promise<number> {
    return await this.dataSource.count();
  }

  async dispose(): Promise<void> {
    return await this.dataSource.dispose();
  }

  async findById(id: string): Promise<T | null> {
    const item = await this.dataSource.findById(id);
    if (!item) return null;

    const metadata = await this.syncMetadataDataSource.findById(
      item.metadataId
    );
    if (!metadata || item.deleted) return null;

    item.metadata = metadata;
    return item;
  }

  async getAll(): Promise<T[]> {
    let items = await this.dataSource.getAll();
    items = items.filter(this._undoRemovedEntities);
    let entities: T[] = [];
    const metadataIds = items.map((item) => item.metadataId);
    const metadatas = await this.syncMetadataDataSource.findByIds(metadataIds);
    const metadataDict = this.createDict(metadatas);
    for (let entity of items) {
      entity.metadata = metadataDict[entity.metadataId];
      entities.push(entity);
    }

    return entities;
  }

  async query(query: (item: T) => boolean): Promise<T[]> {
    const items = await this.dataSource.getAll();
    return items.filter(query);
  }

  async removeMany(entities: T[]): Promise<T[]> {
    const entitiesToSave: T[] = [];
    const metadataTosave: ISyncMetadata[] = [];

    for (let entity of entities) {
      entity.deleted = true;
      entity.metadata.syncOperation = SyncOperationEnum.delete;
      entity.metadata.type = this.type;
      entity = this._dirtyObject(entity);
      entity = this._linkMetadataId(entity, entity.metadata);
      entitiesToSave.push(entity);
      metadataTosave.push(entity.metadata);
    }

    await this.syncMetadataDataSource.updateMany(metadataTosave);
    return await this.dataSource.updateMany(entitiesToSave);
  }

  async remove(entity: T): Promise<T> {
    entity.deleted = true;
    entity.metadata.syncOperation = SyncOperationEnum.delete;
    entity.metadata.type = this.type;
    entity = this._dirtyObject(entity);
    entity = this._linkMetadataId(entity, entity.metadata);
    await this.syncMetadataDataSource.update(
      entity.metadata.id,
      entity.metadata
    );
    this.dataSource.update(entity.id, entity);
    return entity;
  }

  async update(entity: T): Promise<T> {
    entity.metadata.syncOperation = SyncOperationEnum.update;
    entity.metadata.type = this.type;
    entity = this._dirtyObject(entity);
    entity = this._linkMetadataId(entity, entity.metadata);
    await this.syncMetadataDataSource.update(
      entity.metadata.id,
      entity.metadata
    );
    return await this.dataSource.update(entity.id, entity);
  }

  async updateMany(entities: T[]): Promise<T[]> {
    const entitiesToSave: T[] = [];
    const metadataTosave: ISyncMetadata[] = [];

    for (let entity of entities) {
      entity.metadata.syncOperation = SyncOperationEnum.update;
      entity.metadata.type = this.type;
      entity = this._dirtyObject(entity);
      entity = this._linkMetadataId(entity, entity.metadata);
      entitiesToSave.push(entity);
      metadataTosave.push(entity.metadata);
    }

    await this.syncMetadataDataSource.updateMany(metadataTosave);
    return await this.dataSource.updateMany(entitiesToSave);
  }

  async undirtyList(entities: T[]): Promise<T[]> {
    const entitiesToSave: T[] = [];
    const metadataTosave: ISyncMetadata[] = [];

    for (let entity of entities) {
      entity.dirty = false;
      entity = this._linkMetadataId(entity, entity.metadata);
      entitiesToSave.push(entity);
      metadataTosave.push(entity.metadata);
    }

    await this.dataSource.syncUpdate(entitiesToSave);
    await this.syncMetadataDataSource.syncUpdate(metadataTosave);
    return entitiesToSave;
  }

  async hardDelete(): Promise<void> {
    const versionManager = FastSync.getSyncVersionManager();
    await this.dataSource.hardDelete();
    await versionManager.resetTypeSyncVersion(this.type);
  }

  private _undoRemovedEntities(entity: T): boolean {
    return !entity.deleted;
  }

  private _dirtyObject(object: T): T {
    if (!object.dirty) {
      object.dirty = true;
      object.metadata.version++;
    }
    return object;
  }

  private _linkMetadataId(object: T, metadata: ISyncMetadata): T {
    object.metadataId = metadata.id;
    return object;
  }

  private createDict(metadatas: ISyncMetadata[]): {
    [key: string]: ISyncMetadata;
  } {
    const dict: { [key: string]: ISyncMetadata } = {};
    metadatas.forEach((metadata) => {
      dict[metadata.id] = metadata;
    });
    return dict;
  }
}
